driver/src/orderbook/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,341 @@
+use crate::contracts::stablex_auction_element::{StableXAuctionElement, AUCTION_ELEMENT_WIDTH};
+use crate::models::{AccountState, Order};
+use std::collections::HashMap;
+use web3::types::{H160, U256};
+
+/// Handles reading of auction data that has been encoded with the smart
+/// contract's `encodeAuctionElement` function.
+pub struct BatchedAuctionDataReader {
+    /// The account state resulting from unfiltered handled orders.
+    account_state: AccountState,
+    /// All unfiltered orders in the order they were received.
+    orders: Vec<Order>,
+    /// Used when reading data from the smart contract in batches.
+    pagination: Pagination,
+    /// The index of batch whose orders we are filtering for.
+    index: U256,
+    /// The total number of orders per user.
+    user_order_counts: HashMap<H160, usize>,
+}
+
+/// Data for the next call to the smart contract's `getEncodedUsersPaginated`
+/// function (which should be named `getEncodedOrdersPaginated`).
+pub struct Pagination {
+    /// The user of the last received order or H160::zero when no order has been
+    /// received.
+    pub previous_page_user: H160,
+    /// The number of received orders for `previous_page_user`.
+    pub previous_page_user_offset: usize,
+}
+
+impl BatchedAuctionDataReader {
+    /// Create a new BatchedAuctionDataReader.
+    pub fn new(index: U256) -> BatchedAuctionDataReader {
+        BatchedAuctionDataReader {
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+            index,
+            user_order_counts: HashMap::new(),
+        }
+    }
+
+    /// The pagination data used when reading data from the smart contract in
+    /// batches.
+    pub fn pagination(&self) -> &Pagination {
+        &self.pagination
+    }
+
+    /// Signal that no more data will be read and return the result consuming
+    /// self.
+    pub fn get_auction_data(self) -> (AccountState, Vec<Order>) {
+        (self.account_state, self.orders)
+    }
+
+    /// Applies one batch of data.
+    ///
+    /// A batch can come from `getEncodedUsersPaginated` or `getEncodedOrders`.
+    /// In the latter case there is only one batch.
+    ///
+    /// Returns the number of orders in the data, before filtering for `index`.
```

Comment:
nit:  Returns the number of orders in the packed_acution_bytes

I'd also leave out the second half sentence (or change it to before applying *any* filtering) as we are technically filtering by amount as well.